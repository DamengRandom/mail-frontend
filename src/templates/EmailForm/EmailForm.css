.email-form {
  display: flex;
  flex-direction: column;
  max-width: 60%;
  width: 100%;
}

.email-form .email-form__field {
  margin: 10px;
}

.email-form__error {
  text-align: left;
  color: darkred;
  margin-left: 10px;
}

.email-form__input {
  background-color: hsl(0, 0%, 100%);
  border-color: hsl(0, 0%, 80%);
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  cursor: pointer;
  margin: 10px;
  padding: 10px;
  font-size: 16px;
}

textarea.email-form__input {
  font-family: Arial;
}

@media only screen and (max-width: 480px) {
  .email-form {
    max-width: 85%;
  }
}