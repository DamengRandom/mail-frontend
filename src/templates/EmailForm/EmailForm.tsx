// css
import "./EmailForm.css";
import { useState } from "react";
import { useForm, Controller } from "react-hook-form";
import { yupResolver } from '@hookform/resolvers/yup';
// constants
import {
  multiSelectFields,
  baseUrl,
} from "../../constants";
import CreatableSelect from 'react-select/creatable';
// components
import Button from "../../components/Button/Button";
import ErrorMessage from "../../components/ErrorMessage/ErrorMessage";
// helpers
import postData from "../../utils/fetchHelper";
// form validator
import { validationSchema } from "../../utils/formValidator";

type tListType = string[];

export type tFormInputs = {
  provider: tListType,
  from: string,
  to: tListType,
  cc: tListType,
  bcc: tListType,
  subject: string,
  html: string,
}

export type tMultiSelectInput = {
  value: string,
  label: string,
};

export default function EmailForm() {
  const { control, handleSubmit, formState:{ errors } } = useForm({
    resolver: yupResolver(validationSchema),
  });
  const [loader, setLoader] = useState(false);
  const [mailResponse, setMailResponse] = useState({});
  const getEmailInputValue = (array: []) => {
    return array.map(data => (data as tMultiSelectInput).value);
  }

  const onSubmit = (data: tFormInputs) => {
    setLoader(true);
    const {provider, to, cc, bcc, subject, html} = data;
    const requestBody: tFormInputs = {
      provider: getEmailInputValue(provider as []),
      from: 'damon.compliancespace@gmail.com', // @TODO: later can be dynamic (testing only, so made it hardcode first)
      to: getEmailInputValue(to as []),
      cc: getEmailInputValue(cc as []),
      bcc: getEmailInputValue(bcc as []),
      subject,
      html,
    };

    if (provider) {
      provider.map(async (p: any) => {
        await postData(`${baseUrl}/${p.value}`, requestBody)
          .then(data => {
            setMailResponse(data);
            setLoader(false);
          });
      });
    }
  };

  return (
    <>
      <h3>Mailer Service</h3>
      <div className="form-container">
        <form className="email-form" onSubmit={handleSubmit(onSubmit)}>
          {multiSelectFields.map(({ name, placeholder, option }) => (
            <div key={`select-question-${name}`}>
              <Controller
                key={name}
                name={name}
                control={control}
                render={({ field }) => <CreatableSelect
                  key={`select-controller-${field.name}`}
                  className="email-form__field"
                  {...field}
                  isMulti
                  placeholder={placeholder}
                  options={option}
                />}
              />
              {(errors && errors[name])
                && <ErrorMessage
                  className="email-form__error"
                  message={errors[name]?.message}
                />}
            </div>
          ))}
          <Controller
            name="subject"
            control={control}
            defaultValue=""
            render={({ field }) => <input
              className="email-form__field email-form__input"
              placeholder={"Please enter the subject"}
              {...field}
            />}
          />
          {(errors && errors.subject)
            && <ErrorMessage
              className="email-form__error"
              message={errors.subject?.message}
            />}
          <Controller
            name="html"
            control={control}
            defaultValue=""
            render={({ field }) => <textarea
              className="email-form__field email-form__input"
              placeholder={"Please enter the message content"}
              rows={5}
              {...field}
            />}
          />
          {(errors && errors.html)
            && <ErrorMessage
              className="email-form__error"
              message={errors.html?.message}
            />}
          <Button type="submit" text="Submit" disabled={loader} />
        </form>
        {/* context api persist state */}
        <pre>
          {Object.keys(mailResponse).length !== 0 && JSON.stringify(mailResponse, null, 2)}
        </pre>
      </div>
    </>
  );
};
