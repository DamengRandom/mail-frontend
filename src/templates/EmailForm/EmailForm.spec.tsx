// import renderer from 'react-test-renderer';
import { cleanup, render } from '@testing-library/react';
import EmailForm, { tFormInputs } from "./EmailForm";

describe('<EmailForm /> test', () => {
  let componentProps: JSX.IntrinsicAttributes & tFormInputs,
      componentContainer: HTMLElement;

  beforeEach(() => {
    // Arrange
    const props: tFormInputs = {
      provider: ['sendgrid'],
      from: 'test1@test.co',
      to: ['test2@test.co'],
      cc: ['test3@test.co'],
      bcc: ['test4@test.co'],
      subject: 'subject text',
      html: 'body content',
    };
    componentProps = props;
    // Act
    const { container } = render(<EmailForm {...componentProps} />);
    componentContainer = container;
  });

  afterEach(cleanup);

  it('snapshot test for EmailForm component', () => {
    // Assert
    expect(componentContainer.firstChild).toMatchSnapshot();
  });
});
