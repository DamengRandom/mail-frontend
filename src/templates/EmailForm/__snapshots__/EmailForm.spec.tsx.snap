// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmailForm /> test snapshot test for EmailForm component 1`] = `
<h3>
  Mailer Service
</h3>
`;
